### PR TITLE
test: Add Performance test k6 automating script

### PR DESCRIPTION
**Pull request by aluu317**
Signed-off-by: Angel Luu aluu317 [angel.luu@us.ibm.com](angel.luu@us.ibm.com)
Add k6 automating scripts for performance/load test model mesh with test configs, 1 working test template, and payload.
Can be used locally in this first iteration.

### CODE REVIEW

File: perf_test/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+- Different target URLs and/or features.
+
+- Different payload sizes.
+
+## How Howitzer works
+The core design principle beneath Howitzer lies in "exploding" K6 parameter iterables into a set of valid K6 test parameters, rendering each parameter combination into a Javascript object, and injecting the resulting object, along with other test information like the base URL being targeted, into a copy of a K6 template.
+
+When Howitzer is run, the bash script will create directories `render`, `results` and `summary`:
+- `render` stores the "exploded" k6 tests using the k6 template specified in the test config for each k6_opts combination.
+- Then the script will run each individual k6 test in this `render` directory, storing the human readble results as text in `results` directory.
```
Comment (tedhtchang) [inline perf_test/README.md]: readble -> readable

File: perf_test/scripts/renderer.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        if not set(constants.HOWITZER_PARAMS['required']).issubset(set([k for k in test.keys()])):
+            raise KeyError('Missing required test descriptor(s) in test: ' + str(test))
+        if not all(opt in constants.K6_VALID_OPTS.keys() for opt in test['k6_opts']):
+            raise KeyError('One of the following K6 params is unknown: '+ str(test['k6_opts']))
+
+def explode_test(test):
+    """Blows up a single test configuration containing any number of valid iterable
+    params. External parameters, which are the params not directly rendered into
+    a K6 template (i.e. the name of the template) are duplicated across all resulting
+    subtests when a iterable parameter is exploded into multiple K6 cofigurations.
```
Comment (tedhtchang) [inline perf_test/scripts/renderer.py]: cofigurations -> configurations

File: perf_test/scripts/constants.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+"""This file contains the constants related to K6 and Howitzer parameters. The
+large dictionary for K6 options is expecially important because it affects
```
Comment (tedhtchang) [inline perf_test/scripts/constants.py]: expecially -> especially

**Review by tedhtchang — COMMENTED** [review]
There is some typo but works well. thanks.

File: perf_test/scripts/renderer.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
+"""This module explodes the JSON configuration file into a flattened list
```
Comment (joerunde) [inline perf_test/scripts/renderer.py:1]: This file also came with a small set of unit tests as well. Not that this is _super_ important code to test but I am interested in learning how CI works in this gh org if we wanna grab those and set them up to run on merge or something

Comment (aluu317) [reply perf_test/scripts/renderer.py:1]: I copied the unit tests in and can run them locally.. Not sure how the ci works here. I think @tedhtchang mentioned GitHub Action? Ted, could we set something up to run the unit test suite on a branch or something?

Comment (tedhtchang) [reply perf_test/scripts/renderer.py:1]: I tested out in my own repo [here](https://github.com/tedhtchang/kserve-perf/runs/5426470297?check_suite_focus=true)
I think we may add this file to this PR under `.github/workflows/unit-test.yml`:
```
name: k6 Unit Test

on:
  pull_request:
    branches:
      - main
    paths-ignore:
      - 'docs/**'
      - '**.md'
      - '.github/**'
      - 'multi_model_tester/**'

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Install Python 3
        uses: actions/setup-python@v1
        with:
          python-version: 3.9.10
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install -r perf_test/requirements.txt
      - name: Run k6 unit tests
        run: python3 -m unittest unit_test/k6/test_renderer.py
```

Comment (aluu317) [reply perf_test/scripts/renderer.py:1]: Perfect! This works great!

**Comment by tedhtchang** [comment]
/lgtm
thanks @aluu317  @joerunde

**Pull request merged**